src/comm/tcp_server.cpp

```diff
@@ -0,0 +1,316 @@
+
+  // Make read and write ends of pipe nonblocking
+  int flags;
+  flags = fcntl(self_pipe_[0], F_GETFL);
+  if (flags == -1)
+  {
+    throw std::system_error(std::error_code(errno, std::generic_category()), "fcntl-F_GETFL");
+  }
+  flags |= O_NONBLOCK;  // Make read end nonblocking
+  if (fcntl(self_pipe_[0], F_SETFL, flags) == -1)
+  {
+    throw std::system_error(std::error_code(errno, std::generic_category()), "fcntl-F_SETFL");
+  }
+
+  flags = fcntl(self_pipe_[1], F_GETFL);
+  if (flags == -1)
+  {
+    throw std::system_error(std::error_code(errno, std::generic_category()), "fcntl-F_GETFL");
+  }
+  flags |= O_NONBLOCK;  // Make write end nonblocking
+  if (fcntl(self_pipe_[1], F_SETFL, flags) == -1)
+  {
+    throw std::system_error(std::error_code(errno, std::generic_category()), "fcntl-F_SETFL");
+  }
+}
+
+void TCPServer::shutdown()
+{
+  keep_running_ = false;
+
+  // This is basically the self-pipe trick. Writing to the pipe will trigger an event for the event
+  // handler which will stop the select() call from blocking.
+  if (::write(self_pipe_[1], "x", 1) == -1 && errno != EAGAIN)
+  {
+    throw std::system_error(std::error_code(errno, std::generic_category()), "Writing to self-pipe failed.");
+  }
+
+  // After the event loop has finished the thread will be joinable.
+  worker_thread_.join();
+  URCL_LOG_DEBUG("Worker thread joined.");
+}
+
+void TCPServer::bind()
+{
+  struct sockaddr_in server_addr;
+  server_addr.sin_family = AF_INET;
+
+  // INADDR_ANY is a special constant that signalizes "ANY IFACE",
+  server_addr.sin_addr.s_addr = htonl(INADDR_ANY);
+  server_addr.sin_port = htons(port_);
+  int err = ::bind(listen_fd_, (struct sockaddr*)&server_addr, sizeof(server_addr));
+  if (err == -1)
+  {
+    std::ostringstream ss;
+    ss << "Failed to bind socket for port " << port_ << " to address. Reason: " << strerror(errno);
+    throw std::system_error(std::error_code(errno, std::generic_category()), ss.str());
+  }
+  URCL_LOG_DEBUG("Bound %d:%d to FD %d", server_addr.sin_addr.s_addr, port_, (int)listen_fd_);
+
+  FD_SET(listen_fd_, &masterfds_);
+  maxfd_ = std::max((int)listen_fd_, self_pipe_[0]);
+}
+
+void TCPServer::startListen()
+{
+  int err = listen(listen_fd_, 1);
+  if (err == -1)
+  {
+    std::ostringstream ss;
+    ss << "Failed to start listen on port " << port_;
+    throw std::system_error(std::error_code(errno, std::generic_category()), ss.str());
+  }
+  URCL_LOG_DEBUG("Listening on port %d", port_);
+}
+
+void TCPServer::handleConnect()
+{
+  struct sockaddr_storage client_addr;
+  socklen_t addrlen = sizeof(client_addr);
+  int client_fd = accept(listen_fd_, (struct sockaddr*)&client_addr, &addrlen);
+  if (client_fd < 0)
+  {
+    std::ostringstream ss;
+    ss << "Failed to accept connection request on port  " << port_;
+    throw std::system_error(std::error_code(errno, std::generic_category()), ss.str());
+  }
+
+  if (client_fds_.size() < max_clients_allowed_)
```

Comment:
There needs to be a check on whether max_client_allowed_ is equal to 0. In order for 0 to be an unlimited number of clients allowed.

Reply:
Yes, absolutely! I've got the script_sender restricted locally during my tests... I'll fix that.

Reply:
Fixed in   412640939c547b990ad383995bce9a9a411b6c42